### DISCUSSION TITLE
Use Conventional Commits and semantic-release

### DISCUSSION THREAD

**Discussion by kodebach** [ElektraInitiative/libelektra#3740] [Ideas]
@markus2330 @mpranj @robaerd

Did any of you consider using "Conventional Commits" and the semantic-release system? It would semi-automate the creation of the release notes and reduce merge conflicts (because the `_preparation_next_release.md` isn't needed anymore).

https://www.conventionalcommits.org/en/v1.0.0/
https://semantic-release.gitbook.io/semantic-release/

The commit format could be checked with a GitHub action, similar to what we do for the release notes. The check should, however, be separate so that we can develop code without using the convention for every commit. Then at the end of a PR we squash everything down into one (or a few) commit that conform to the convention. Always using a squash merge (AFAIK this can be enforced in the GitHub settings) would also be an option. But then you have to remember to write a correct description when merging, because the commit message for a squash merge cannot be checked AFAIK.

There are some pre-built actions that do checks similar to "Conventional Commits" for PRs (e.g. https://github.com/amannn/action-semantic-pull-request). We could also build something custom. Then we could enforce that the PR description follows "Conventional Commits". I think the PR description is the default suggested by GitHub when using a squash merge.

---

**Note:** This is only a vague idea, but I think something like this would be much better than our current `_preparation_next_release.md`. If we do anything like this, it should only be after 1.0. Then PRs should become smaller and which is much more manageable for "Conventional Commits".

**Comment by markus2330** [comment]
I tried to do something like this (long time ago) but it is too difficult and the result is not as wanted. The wanted granularity of (merge) commits and release notes simply does not match.

Furthermore, it would be a huge practical problem as the commit message of the merge commits are written by the merger and we want the release notes written by the authors.

The constant merge conflicts are a problem of GitHub, not from Git. If you merge locally you see that there are no merge conflicts due the union merging. (doc/news/.gitattributes)

Would be interesting if others (like GitLab) are handling this better. Recently GitHub made problems that I could not add someone and needed a reverification that I am employed at an university. So long-term we probably need to move away from GitHub anyway.

**Reply by mpranj** [reply]
> So long-term we probably need to move away from GitHub anyway.

In #3603 we discussed moving some of our CI to Github Actions. I assume we should not migrate any further jobs until GitHub vs GitLab is decided?

**Reply by kodebach** [reply]
> but it is too difficult and the result is not as wanted. The wanted granularity of (merge) commits and release notes simply does not match.

Yes, adhering strictly to "1 change per commit" or "1 change per PR" can be quite hard. Directly using the specs I linked to may not be a good solution, but I still think something similar would be an improvement to our current flow. I suggested an alternative below.

> commit message of the merge commits are written by the merger and we want the release notes written by the authors.

Semantic release uses the messages of **all** commits not just merge commits. In fact AFAIK normally merge commits are ignored (since they won't conform to conventional commits).

> If you merge locally you see that there are no merge conflicts due the union merging.

Union merging avoids the conflict, but doesn't necessarily produce correct release notes.

> Recently GitHub made problems that I could not add someone and needed a reverification that I am employed at an university

AFAIK this verification is only needed to receive the GitHub Enterprise Features for free. But some time ago GitHub made both private and public repos with unlimited collaborators [completely free](https://github.com/pricing) (I think this was after the acquisition by Microsoft).

> So long-term we probably need to move away from GitHub anyway.

IMO the only other option would be moving to gitlab.com (**not** the self-hosted version). Any self-hosted git repo would be to much maintenance effort. I don't think we have the capacity to run our own git server. Even the PHP team is [moving to GitHub](https://www.theregister.com/2021/03/29/php_repository_infected/) after some security problems.

Also from a CI perspective Gitlab may have more features, but we don't use any of the advanced features and using Gitlab would create a few problems. 1) Cirrus CI has no native support for Gitlab, but there is a [3rd party workaround](https://github.com/sio/cirrus-run) (oddly enough this is hosted on GitHub). 2) Gitlab CI only provides 400 Minutes per group (could have multiple projects) per month. We would use that up very quickly. We could use self-hosted runners, but GitHub Actions allows that too.

**Reply by markus2330** [reply]
> AFAIK this verification is only needed to receive the GitHub Enterprise Features for free.

Unfortunately, having more than 40 people in a group, what we have, already seems to be an "Enterprise" feature. Fact is that I could not add someone until I did the reverification. I asked for clarification but didn't get an answer yet. Let us wait for this before going into any details.

**Reply by kodebach** [reply]
I couldn't find any info about any limit on the size of an organization for GitHub Free. In fact the docs say there is a limit for paid orgs but not for free ones. If you use a paid tier you obviously only get the number of users you pay for.

Possibly the error is because the ElektraInitiative organization was created before free organizations were a thing? Maybe it is now somehow marked as an Enterprise org and you need Enterprise status to manage it? Maybe you can downgrade to the Free tier somewhere?

I had some problems myself, when I went from Education status to Free tier (don't need the Education stuff). I remember having to fiddle with some settings to make a "You don't have a payment method configured" error go away. So maybe this is similar and GitHub tried to convert your account from Education to Enterprise and if you manually go to Free tier it works fine. But if it works now, it's probably best to wait for the clarification.

**Reply by markus2330** [reply]
I suspect that "Free tier" has some hidden (?) serious limits (like GB and #seats) because they said now: "Hi there,

We also have nonprofit discounts as an option. If you qualify, you can apply for a nonprofit discount by following these directions:

https://github.com/nonprofit

After following those instructions and applying, our team will then reach out to you about your potential discount status.

If you run into any issues, just let us know.

Thanks,

Nik
GitHub Support"

**Reply by kodebach** [reply]
There definitely are limits for the Free tier and they are [listed on the pricing site](https://github.com/pricing#compare). But all the limits that would matter to us only apply to private repositories. If there were any hidden limits I'm sure we would find more information about them on Google...

My guess is still that GitHub put your account into some kind of grace period status until you re-verified, switch to a payed plan or downgraded to the free tier. And because of that status some features didn't work correctly.

**Reply by markus2330** [reply]
The "problem" could be the "abgaben" repo, which is private and the whole ElektraInitiative group has access to it.

**Reply by kodebach** [reply]
I randomly found this post today: https://github.community/t/what-do-these-changes-mean-for-me-if-i-m-on-a-legacy-plan/18547

Could it be that the Elektra organisation is still on one of those legacy plans and that's why there were problems adding people without the education status?

**Reply by markus2330** [reply]
I don't know on which plan we are. As we renewed quite recently we are probably not on legacy plans?

**Reply by kodebach** [reply]
[These](https://docs.github.com/en/billing/managing-billing-for-your-github-account/upgrading-your-github-subscription#upgrading-your-organizations-subscription) are the steps for changing the subscription of an organization, I assume somewhere on that settings page, it should list the current plan.

> As we renewed quite recently we are probably not on legacy plans?

I'm not sure about that. The new plans are all either free or per-user pricing. Maybe confirming your university status, just extended the legacy plan. It doesn't really matter either way right now, I just wanted to mention that it since I saw the post.

**Reply by markus2330** [reply]
The current plan is: "GitHub Team
Advanced collaboration and support for organizations"

**Comment by kodebach** [comment]
I thought about a few alternatives, but I think the best solution would be this:

- Every PR `#XXX` **must** add a plain-text file `doc/news/next_release/#XXX`.
- For each change the PR makes (e.g. bugfix, new feature, doc change, etc.) there has to be a description in the file.
- Change descriptions are separated by `---` (alone in a line).
- Every change description on its own follows the conventional commit specification.
- We also require an `Author` footer for each description (like the `Reviewed-by` [here](https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-multi-paragraph-body-and-multiple-footers))
- For things like "Highlights" we could also use footers.
- To generate release notes we basically do the same as [semantic-release](https://semantic-release.gitbook.io/semantic-release/), but instead of using commit messages we use our `doc/news/next_release/#XXX` files.

This way we still get the granularity we want (n changes per PR) and the difficulty is basically the same as the current system. But crucially, since we use one file per PR there shouldn't be any merge conflicts (not locally and not on GitHub). If a PR is an update to a previous one, we could also allow adding e.g. `doc/news/next_release/#XXX.update.#YYY` to indicate `#XXX` doesn't have its own file but instead update the file of `#YYY`.

---

As a reminder, these are just ideas and I would not implement them before 1.0 (unless somebody has the time to test all this).

**Reply by markus2330** [reply]
Having several files is basically a workaround for the non-working union merge driver. Let us wait for the GitHub answer and if we switch we would see if it is also a problem in GitLab.

> As a reminder, these are just ideas and I would not implement them before 1.0 (unless somebody has the time to test all this).

I don't see this related to any milestone.

**Reply by kodebach** [reply]
> I don't see this related to any milestone.

It depends on the solution we go with. Before 1.0 there will be some big PRs that will be hard to split neatly into conventional commit messages. After 1.0 everything should be stable and changes should be a lot more separate from each other.
